File: server/services/agent/supervisor.ts
import { GoogleGenAI, Part, Content } from "@google/genai";
import { sessionService } from './state';
import { toolRegistry } from './registry';
import { toolManagerService } from '../tools/manager';
import { SUPERVISOR_SYSTEM } from '../prompts';
import { GEMINI_API_KEY } from '../../config';
import { AgentMode, Industry, AgentRoleConfig, ThinkingStep, Plan, ChartData } from '../../../types';

// Helper types
interface SupervisorAgentOptions {
    industry?: Industry | string;
    role?: AgentRoleConfig;
    customRules?: string;
    mode?: AgentMode;
}

// Retry logic
async function retryWithBackoff<T>(fn: () => Promise<T>, maxRetries = 3): Promise<T> {
  let delay = 1000;
  for (let i = 0; i < maxRetries; i++) {
    try {
      return await fn();
    } catch (error: any) {
      if ((error?.status === 429 || error?.message?.includes("429")) && i < maxRetries - 1) {
        await new Promise((res) => setTimeout(res, delay));
        delay *= 2;
        continue;
      }
      throw error;
    }
  }
  return await fn();
}

// Tool Helpers
function toolLabel(name: string): string {
    const tool = toolRegistry.get(name);
    const desc = tool?.definition?.description as string | undefined;
    if (!desc) return name;
    const first = desc.split(/[。；\.;]/)[0]?.trim();
    if (!first) return name;
    return first.length > 36 ? first.slice(0, 33) + "..." : first;
}

function thoughtSummary(raw: string, maxLen = 120): string {
    const trimmed = raw.trim();
    if (!trimmed) return "";
    const firstLine = trimmed.split(/\n/)[0]?.trim() ?? trimmed;
    return firstLine.length > maxLen ? firstLine.slice(0, maxLen) + "..." : firstLine;
}

// Execute Tool Calls
async function executeToolCalls(
    functionCalls: Array<{ name: string; id?: string; args: Record<string, unknown> }>,
    sessionId: string,
    blockingIds: Set<string>,
    callbacks: {
        onThinking: (step: ThinkingStep) => void;
        onPlanProposed?: (plan: Plan) => void;
        onChartData?: (data: ChartData) => void;
        onPlanStepUpdate?: (stepId: string, status: "in_progress" | "completed") => void;
    }
): Promise<{ parts: Part[]; writtenFilePaths: string[] }> {
    const hasPlanCall = functionCalls.some((f) => f.name === "propose_plan");
    const writtenFilePaths: string[] = [];
    const parts: Part[] = [];

    const executeOne = async (call: (typeof functionCalls)[0]) => {
        const stepId = `call-${call.id ?? call.name}`;
        const label = toolLabel(call.name);

        // Block execution if plan is pending approval (except propose_plan itself)
        if (hasPlanCall && call.name !== "propose_plan") {
            return {
                functionResponse: {
                    name: call.name,
                    response: { error: "Execution blocked: Plan must be approved first." },
                    id: call.id,
                },
            } as Part;
        }

        if (call.name === "generate_chart") {
            callbacks.onChartData?.(call.args as unknown as ChartData);
            return {
                functionResponse: { name: call.name, response: { status: "CHART_RENDERED" }, id: call.id },
            } as Part;
        }

        const isBlocking = blockingIds.has(call.name);
        // write_file and generate_chart don't show "Thinking" for start, mostly result
        const isWriteFile = call.name === "write_file";
        const isGenerateChart = call.name === "generate_chart";

        if (!isWriteFile && !isGenerateChart) {
            callbacks.onThinking({
                id: stepId,
                agentId: call.name,
                agentName: label,
                content: isBlocking ? `${label}（子代理执行中...）` : label,
                status: "active",
                timestamp: Date.now(),
            });
        }

        try {
            const result = await toolRegistry.execute(
                call.name,
                call.args,
                sessionId,
                (p) => {
                    if (!isWriteFile && !isGenerateChart)
                        callbacks.onThinking({
                            id: stepId,
                            agentId: call.name,
                            agentName: label,
                            content: p,
                            status: "active",
                            timestamp: Date.now(),
                        });
                }
            );

            if (call.name === "report_step_done") {
                const stepId = (call.args as { stepId?: string })?.stepId;
                if (stepId) callbacks.onPlanStepUpdate?.(stepId, "completed");
            }

            if (call.name === "propose_plan" && typeof result === "object" && result !== null && "plan" in result) {
                callbacks.onPlanProposed?.({ ...(result as any).plan, isApproved: false, isCollapsed: false });
            }

            if (isWriteFile && typeof result === "object" && result !== null && "path" in result) {
                writtenFilePaths.push((result as any).path);
            }

            if (!isWriteFile && !isGenerateChart) {
                const doneContent =
                    typeof result === "object" && result !== null && "plan" in result
                        ? `Plan: ${(result as any).plan?.title ?? "执行计划"}`
                        : typeof result === "object" && result !== null && "analysis" in result
                        ? (result as any).analysis?.slice?.(0, 80) + ((result as any).analysis?.length > 80 ? "..." : "") || label
                        : label;

                callbacks.onThinking({
                    id: stepId,
                    agentId: call.name,
                    agentName: label,
                    content: doneContent,
                    details: JSON.stringify(result, null, 2),
                    status: "completed",
                    timestamp: Date.now(),
                });
            }
            return {
                functionResponse: { name: call.name, response: { result }, id: call.id },
            } as Part;
        } catch (e: any) {
             if (!isWriteFile && !isGenerateChart)
                callbacks.onThinking({
                    id: stepId,
                    agentId: call.name,
                    agentName: label,
                    content: `Failed: ${e.message}`,
                    details: e.message,
                    status: "failed",
                    timestamp: Date.now(),
                });
            return {
                functionResponse: { name: call.name, response: { error: e.message }, id: call.id },
            } as Part;
        }
    };

    const blockingCalls = functionCalls.filter((f) => blockingIds.has(f.name));
    const nonBlockingCalls = functionCalls.filter((f) => !blockingIds.has(f.name));

    const blockingResponses = await Promise.all(blockingCalls.map(executeOne));
    const nonBlockingResponses = await Promise.all(nonBlockingCalls.map(executeOne));

    parts.push(...blockingResponses, ...nonBlockingResponses);
    return { parts, writtenFilePaths };
}


// Supervisor Agent Loop
export const supervisorAgent = async (
    sessionId: string,
    prompt: string,
    callbacks: {
        onThinking: (step: ThinkingStep) => void;
        onText: (chunk: string) => void;
        onPlanProposed?: (plan: Plan) => void;
        onChartData?: (data: ChartData) => void;
        onFilesWritten?: (paths: string[]) => void;
        onPlanStepUpdate?: (msgId: string, stepId: string, status: "in_progress" | "completed") => void;
    },
    params: {
        resumePlan?: Plan;
        isApprovalConfirmed?: boolean;
        planMsgId?: string;
        options?: SupervisorAgentOptions;
    }
) => {
    if (!GEMINI_API_KEY) throw new Error("GEMINI_API_KEY not configured");
    const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });
    
    const sessionState = await sessionService.getSession(sessionId);
    let history: Content[] = sessionState?.geminiHistory || [];
    const blockingIds = toolRegistry.getBlockingIds();

    const mode = params.options?.mode ?? AgentMode.AGENTIC;
    
    // Construct System Instruction
    const instruction = SUPERVISOR_SYSTEM({
        industry: params.options?.industry as Industry, // Cast or handle type mismatch
        role: params.options?.role,
        customRules: params.options?.customRules,
        mode,
        hasMcpConnected: false // Backend: handle MCP separately if needed
    });

    // Filter tools
    const toolIds = toolRegistry.getDefinitions().map(d => d.name);
    let toolDefinitions = toolRegistry.getDefinitions()
        .filter((t) => t.name && toolManagerService.getToolEnabled(t.name));
        
    if (mode === AgentMode.TRADITIONAL) {
        toolDefinitions = toolDefinitions.filter((t) => t.name !== "propose_plan");
    }

    const thinkingBudget = mode === AgentMode.DEEP_SEARCH ? 16000 : 8000;

    // Handle User Input / Resume Plan
    if (prompt || params.resumePlan) {
        let userText = prompt;
        if (params.resumePlan && params.isApprovalConfirmed) {
            userText = `计划已批准。请按计划立即开始执行所有步骤：${JSON.stringify(params.resumePlan.steps.filter((s) => s.approved))}`;
        }
        history.push({ role: "user", parts: [{ text: userText }] });
    }

    let loopCount = 0;
    let currentTurnText = "";
    let thoughtStepId: string | null = null;

    while (loopCount < 10) {
        loopCount++;
        const turnId = `turn-${loopCount}`;

        const stream = await retryWithBackoff(() =>
            ai.models.generateContentStream({
                model: "gemini-2.0-flash-thinking-exp-01-21", // Using thinking model
                contents: history,
                config: {
                    systemInstruction: instruction,
                    tools: [{ functionDeclarations: toolDefinitions }],
                    // thinkingConfig: { thinkingBudget }, // Sdk param check needed
                },
            })
        );

        const accumulatedParts: any[] = [];
        currentTurnText = "";
        let currentThought = "";
        thoughtStepId = `th-${turnId}`;

        for await (const chunk of stream) {
            const candidate = chunk.candidates?.[0];
            if (candidate?.content?.parts) {
                for (const part of candidate.content.parts) {
                    if (part.text) {
                        // Close thought if open
                        if (thoughtStepId && currentThought.trim()) {
                            callbacks.onThinking({
                                id: thoughtStepId,
                                agentId: "supervisor",
                                agentName: "Supervisor",
                                content: thoughtSummary(currentThought),
                                details: currentThought,
                                status: "completed",
                                timestamp: Date.now(),
                            });
                            thoughtStepId = null;
                        }

                        currentTurnText += part.text;
                        callbacks.onText(currentTurnText);
                        
                        if (accumulatedParts.length > 0 && accumulatedParts[accumulatedParts.length - 1].text) {
                            accumulatedParts[accumulatedParts.length - 1].text += part.text;
                        } else {
                            accumulatedParts.push({ text: part.text });
                        }
                    } else if ((part as any).thought) { // Check SDK type for thought
                        currentThought += (part as any).thought;
                        const summary = thoughtSummary(currentThought);
                        if (summary) {
                            callbacks.onThinking({
                                id: thoughtStepId!,
                                agentId: "supervisor",
                                agentName: "Supervisor",
                                content: summary,
                                details: currentThought,
                                status: "active",
                                timestamp: Date.now(),
                            });
                        }
                        accumulatedParts.push(part);
                    } else if (part.functionCall) {
                         // Close thought if open
                        if (thoughtStepId && currentThought.trim()) {
                            callbacks.onThinking({
                                id: thoughtStepId,
                                agentId: "supervisor",
                                agentName: "Supervisor",
                                content: thoughtSummary(currentThought),
                                details: currentThought,
                                status: "completed",
                                timestamp: Date.now(),
                            });
                            thoughtStepId = null;
                        }
                        
                        const fc = part.functionCall as any;
                        const fcId = fc.id ?? `fc-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
                        const fcWithId = { ...fc, id: fcId };
                        
                        if (fc.name !== "write_file" && fc.name !== "generate_chart") {
                            const stepId = `call-${fcId}`;
                            const label = toolLabel(fc.name);
                            callbacks.onThinking({
                                id: stepId,
                                agentId: fc.name,
                                agentName: label,
                                content: label,
                                status: "pending",
                                timestamp: Date.now(),
                            });
                        }
                        accumulatedParts.push({ ...part, functionCall: fcWithId });
                    }
                }
            }
        }

        // Final thought close
        if (thoughtStepId && currentThought.trim()) {
             callbacks.onThinking({
                id: thoughtStepId,
                agentId: "supervisor",
                agentName: "Supervisor",
                content: thoughtSummary(currentThought),
                details: currentThought,
                status: "completed",
                timestamp: Date.now(),
            });
            thoughtStepId = null;
        }

        history.push({ role: "model", parts: accumulatedParts });
        
        // Execute Tools
        const functionCalls = accumulatedParts.filter((p) => p.functionCall).map((p) => p.functionCall);

        if (functionCalls.length > 0) {
            const hasPlanCall = functionCalls.some((f: any) => f.name === "propose_plan");
            const blockingCalls = functionCalls.filter((f: any) => blockingIds.has(f.name));

            const { parts: responseParts, writtenFilePaths } = await executeToolCalls(
                functionCalls,
                sessionId,
                blockingIds,
                {
                    onThinking: callbacks.onThinking,
                    onPlanProposed: callbacks.onPlanProposed,
                    onChartData: callbacks.onChartData,
                    onPlanStepUpdate: params.planMsgId && callbacks.onPlanStepUpdate
                        ? (stepId, status) => callbacks.onPlanStepUpdate!(params.planMsgId!, stepId, status)
                        : undefined,
                }
            );

            if (writtenFilePaths.length > 0 && callbacks.onFilesWritten) {
                callbacks.onFilesWritten(writtenFilePaths);
            }

            history.push({ role: "user", parts: responseParts });

            if (hasPlanCall) break;
            if (blockingCalls.length > 0 && !hasPlanCall) continue;
        } else {
            break;
        }
    }

    await sessionService.updateSession(sessionId, { geminiHistory: history });
    return { text: currentTurnText };
};

File: server/internal/agent/supervisor.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"math/rand"
	"strings"
	"time"

	"agentic-demo/server/internal/config"
	"agentic-demo/server/internal/prompts"
	"agentic-demo/server/internal/registry"
	"agentic-demo/server/internal/registry/builtin"
	"agentic-demo/server/internal/store"

	"google.golang.org/genai"
)

type SupervisorOptions struct {
	Industry string
	Mode     string
}

type Callbacks struct {
	OnThinking       func(step map[string]any)
	OnText           func(content string)
	OnPlanProposed   func(plan map[string]any)
	OnChartData      func(data map[string]any)
	OnFilesWritten   func(paths []string)
	OnPlanStepUpdate func(msgID, stepID, status string)
	OnToast          func(msg string)
}

type SupervisorDeps struct {
	Store        *store.SessionStore
	TodoStore    *store.TodoStore
	Registry     *registry.Registry
	ToolEnable   *store.ToolEnableStore // optional: filter tools by enabled state
}

func RunSupervisor(
	ctx context.Context,
	deps SupervisorDeps,
	sessionID, message string,
	opts SupervisorOptions,
	cb Callbacks,
	params *struct {
		ResumePlan          interface{}
		IsApprovalConfirmed bool
		PlanMsgID           string
	},
) error {
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  config.GeminiAPIKey,
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		return fmt.Errorf("create genai client: %w", err)
	}
	defer client.Close()

	session, err := deps.Store.GetSession(sessionID)
	if err != nil || session == nil {
		session = &store.AgentSessionState{
			SessionID:      sessionID,
			Title:          "新会话",
			GeminiHistory:  []any{},
			UIMessages:     []any{},
			VFS:            store.DefaultVFS(),
			KnowledgeChunks: []store.KnowledgeChunk{},
		}
		_ = deps.Store.SaveSession(sessionID, session)
	}

	history := session.GeminiHistory
	if history == nil {
		history = []any{}
	}

	userText := message
	if params != nil && params.IsApprovalConfirmed && params.ResumePlan != nil {
		userText = "计划已批准。请按计划立即开始执行。"
	}

	history = append(history, map[string]any{
		"role": "user",
		"parts": []any{map[string]string{"text": userText}},
	})

	systemInstruction := prompts.SupervisorSystem(prompts.SupervisorVars{
		Industry:        opts.Industry,
		Mode:            opts.Mode,
		HasMcpConnected: false,
	})

	var toolDefs []*genai.FunctionDeclaration
	if deps.ToolEnable != nil {
		toolDefs = deps.Registry.GetDefinitionsEnabled(deps.ToolEnable.GetEnabled)
	} else {
		toolDefs = deps.Registry.GetDefinitions()
	}
	blockingIDs := deps.Registry.GetBlockingIDs()

	userMsgID := fmt.Sprintf("u-%d", time.Now().UnixMilli())
	assistantMsgID := fmt.Sprintf("agent-%d", time.Now().UnixMilli())

	existingUIMessages := sliceAny(session.UIMessages)
	uiMessages := append(existingUIMessages, map[string]any{
		"id": userMsgID, "role": "user", "content": userText, "timestamp": time.Now().UnixMilli(),
	}, map[string]any{
		"id": assistantMsgID, "role": "assistant", "content": "", "thinkingSteps": []any{}, "timestamp": time.Now().UnixMilli(),
	})

	cfg := &genai.GenerateContentConfig{
		SystemInstruction: &genai.Content{Parts: []*genai.Part{{Text: systemInstruction}}},
		Tools:             []*genai.Tool{{FunctionDeclarations: toolDefs}},
	}

	loopCount := 0
	currentTurnText := ""
	for loopCount < 10 {
		loopCount++
		contents := historyToContents(history)
		currentTurnText = ""
		var accumulatedParts []*genai.Part
		currentThought := ""
		thoughtStepID := fmt.Sprintf("th-%d", loopCount)
		writtenFilePaths := []string{}
		charts := []map[string]any{}
		thinkingSteps := []map[string]any{}

		for resp, err := range client.Models.GenerateContentStream(ctx, "gemini-2.0-flash", contents, cfg) {
			if err != nil {
				return err
			}
			if resp == nil || len(resp.Candidates) == 0 {
				continue
			}
			cand := resp.Candidates[0]
			if cand.Content == nil || len(cand.Content.Parts) == 0 {
				continue
			}
			for _, p := range cand.Content.Parts {
				if p.Text != "" {
					if currentThought != "" {
						if cb.OnThinking != nil {
							cb.OnThinking(thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
						}
						thinkingSteps = append(thinkingSteps, thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
						accumulatedParts = append(accumulatedParts, &genai.Part{Thought: currentThought})
						currentThought = ""
					}
					currentTurnText += p.Text
					if cb.OnText != nil {
						cb.OnText(currentTurnText)
					}
					if len(accumulatedParts) > 0 && accumulatedParts[len(accumulatedParts)-1].Text != "" {
						accumulatedParts[len(accumulatedParts)-1].Text += p.Text
					} else {
						accumulatedParts = append(accumulatedParts, &genai.Part{Text: p.Text})
					}
				} else if p.Thought != "" {
					currentThought += p.Thought
					if cb.OnThinking != nil {
						cb.OnThinking(thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "active", currentThought))
					}
					thinkingSteps = appendOrUpdate(thinkingSteps, thoughtStepID, thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "active", currentThought))
					accumulatedParts = append(accumulatedParts, &genai.Part{Thought: p.Thought})
				} else if p.FunctionCall != nil {
					if currentThought != "" {
						accumulatedParts = append(accumulatedParts, &genai.Part{Thought: currentThought})
						if cb.OnThinking != nil {
							cb.OnThinking(thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
						}
						thinkingSteps = appendOrUpdate(thinkingSteps, thoughtStepID, thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
						currentThought = ""
					}
					fc := p.FunctionCall
					fcID := fc.Id
					if fcID == "" {
						fcID = fmt.Sprintf("fc-%d-%x", time.Now().UnixMilli(), rand.Uint32())
						fc = &genai.FunctionCall{Name: fc.Name, Id: fcID, Args: fc.Args}
					}
					if fc.Name != "write_file" && fc.Name != "generate_chart" {
						label := toolLabel(fc.Name)
						if cb.OnThinking != nil {
							cb.OnThinking(thinkingStep("call-"+fcID, fc.Name, label, label, "pending", ""))
						}
						thinkingSteps = append(thinkingSteps, thinkingStep("call-"+fcID, fc.Name, label, label, "pending", ""))
					}
					accumulatedParts = append(accumulatedParts, &genai.Part{FunctionCall: fc})
				}
			}
		}

		if currentThought != "" {
			if cb.OnThinking != nil {
				cb.OnThinking(thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
			}
			thinkingSteps = appendOrUpdate(thinkingSteps, thoughtStepID, thinkingStep(thoughtStepID, "supervisor", "Supervisor", thoughtSummary(currentThought, 120), "completed", currentThought))
		}

		modelPartsForHistory := partsToHistory(accumulatedParts)
		history = append(history, map[string]any{
			"role":  "model",
			"parts": modelPartsForHistory,
		})

		var functionCalls []struct {
			Name string
			Id   string
			Args json.RawMessage
		}
		for _, p := range accumulatedParts {
			if p.FunctionCall != nil {
				functionCalls = append(functionCalls, struct {
					Name string
					Id   string
					Args json.RawMessage
				}{Name: p.FunctionCall.Name, Id: p.FunctionCall.Id, Args: p.FunctionCall.Args})
			}
		}

		if len(functionCalls) == 0 {
			break
		}

		hasPlanCall := false
		for _, fc := range functionCalls {
			if fc.Name == "propose_plan" {
				hasPlanCall = true
				break
			}
		}

		var responseParts []*genai.Part
		execReq := registry.ExecuteRequest{
			Ctx:          ctx,
			SessionID:    sessionID,
			Store:        deps.Store,
			TodoStore:    deps.TodoStore,
			GeminiClient: client,
			OnProgress:   nil,
		}

		blockingCalls := []struct {
			Name string
			Id   string
			Args json.RawMessage
		}{}
		nonBlockingCalls := []struct {
			Name string
			Id   string
			Args json.RawMessage
		}{}
		for _, fc := range functionCalls {
			if blockingIDs[fc.Name] {
				blockingCalls = append(blockingCalls, fc)
			} else {
				nonBlockingCalls = append(nonBlockingCalls, fc)
			}
		}

		for _, fc := range append(blockingCalls, nonBlockingCalls...) {
			if hasPlanCall && fc.Name != "propose_plan" {
				responseParts = append(responseParts, &genai.Part{
					FunctionResponse: &genai.FunctionResponse{
						Name: fc.Name,
						Id:   fc.Id,
						Response: map[string]any{
							"error": "Execution blocked: Plan must be approved first.",
						},
					},
				})
				continue
			}

			if fc.Name == "generate_chart" {
				var chartArgs map[string]any
				_ = json.Unmarshal(fc.Args, &chartArgs)
				if cb.OnChartData != nil {
					cb.OnChartData(chartArgs)
				}
				charts = append(charts, chartArgs)
				responseParts = append(responseParts, &genai.Part{
					FunctionResponse: &genai.FunctionResponse{Name: fc.Name, Id: fc.Id, Response: map[string]any{"status": "CHART_RENDERED"}},
				})
				continue
			}

			result, execErr := deps.Registry.Execute(execReq, fc.Name, fc.Args)
			stepID := "call-" + fc.Id
			label := toolLabel(fc.Name)

			if execErr != nil {
				if cb.OnThinking != nil {
					cb.OnThinking(thinkingStep(stepID, fc.Name, label, "Failed: "+execErr.Error(), "failed", execErr.Error()))
				}
				thinkingSteps = appendOrUpdate(thinkingSteps, stepID, thinkingStep(stepID, fc.Name, label, "Failed: "+execErr.Error(), "failed", execErr.Error()))
				responseParts = append(responseParts, &genai.Part{
					FunctionResponse: &genai.FunctionResponse{Name: fc.Name, Id: fc.Id, Response: map[string]any{"error": execErr.Error()}},
				})
				continue
			}

			if fc.Name == "report_step_done" && params != nil && params.PlanMsgID != "" && cb.OnPlanStepUpdate != nil {
				var res struct {
					StepID string `json:"stepId"`
				}
				if m, ok := result.(map[string]interface{}); ok {
					if sid, ok := m["stepId"].(string); ok {
						cb.OnPlanStepUpdate(params.PlanMsgID, sid, "completed")
					}
				}
				_ = res
			}

			if fc.Name == "propose_plan" {
				if m, ok := result.(map[string]interface{}); ok {
					if plan, ok := m["plan"].(map[string]interface{}); ok {
						plan["isApproved"] = false
						if cb.OnPlanProposed != nil {
							cb.OnPlanProposed(plan)
						}
					}
				}
			}

			if fc.Name == "write_file" {
				if m, ok := result.(map[string]interface{}); ok {
					if path, ok := m["path"].(string); ok {
						writtenFilePaths = append(writtenFilePaths, path)
						if cb.OnFilesWritten != nil {
							cb.OnFilesWritten(writtenFilePaths)
						}
					}
				}
			}

			if fc.Name != "write_file" && fc.Name != "generate_chart" {
				doneContent := label
				if m, ok := result.(map[string]interface{}); ok {
					if p, hasPlan := m["plan"]; hasPlan {
						if pm, ok := p.(map[string]interface{}); ok {
							if t, ok := pm["title"].(string); ok {
								doneContent = "Plan: " + t
							}
						}
					} else if a, hasAnalysis := m["analysis"]; hasAnalysis {
						if as, ok := a.(string); ok && len(as) > 80 {
							doneContent = as[:80] + "..."
						} else if as, ok := a.(string); ok {
							doneContent = as
						}
					}
				}
				if cb.OnThinking != nil {
					cb.OnThinking(thinkingStep(stepID, fc.Name, label, doneContent, "completed", fmt.Sprintf("%v", result)))
				}
				thinkingSteps = appendOrUpdate(thinkingSteps, stepID, thinkingStep(stepID, fc.Name, label, doneContent, "completed", fmt.Sprintf("%v", result)))
			}

			responseParts = append(responseParts, &genai.Part{
				FunctionResponse: &genai.FunctionResponse{Name: fc.Name, Id: fc.Id, Response: map[string]any{"result": result}},
			})
		}

		history = append(history, map[string]any{
			"role":  "user",
			"parts": functionResponsesToHistory(responseParts),
		})

		if hasPlanCall {
			break
		}
		if len(blockingCalls) > 0 {
			continue
		}
	}

	assistantMsg := map[string]any{
		"id":            assistantMsgID,
		"role":          "assistant",
		"content":       currentTurnText,
		"thinkingSteps": thinkingSteps,
		"timestamp":     time.Now().UnixMilli(),
	}
	if len(charts) > 0 {
		assistantMsg["charts"] = charts
	}
	if len(writtenFilePaths) > 0 {
		assistantMsg["writtenFiles"] = writtenFilePaths
	}

	finalUIMessages := make([]any, 0, len(uiMessages))
	replaced := false
	for _, m := range uiMessages {
		mm, ok := m.(map[string]any)
		if !ok {
			finalUIMessages = append(finalUIMessages, m)
			continue
		}
		if rid, _ := mm["id"].(string); rid == assistantMsgID {
			finalUIMessages = append(finalUIMessages, assistantMsg)
			replaced = true
		} else {
			finalUIMessages = append(finalUIMessages, mm)
		}
	}
	if !replaced {
		finalUIMessages = append(finalUIMessages, assistantMsg)
	}

	_ = deps.Store.UpdateSession(sessionID, map[string]any{
		"geminiHistory": history,
		"uiMessages":    finalUIMessages,
	})
	return nil
}

func gatherThinkingSteps(uiMessages []any, steps []map[string]any) []any {
	seen := make(map[string]bool)
	var out []any
	for _, s := range steps {
		id, _ := s["id"].(string)
		if !seen[id] {
			seen[id] = true
			out = append(out, s)
		}
	}
	return out
}

func partsToHistory(parts []*genai.Part) []any {
	var out []any
	for _, p := range parts {
		if p.Text != "" {
			out = append(out, map[string]string{"text": p.Text})
		} else if p.Thought != "" {
			out = append(out, map[string]any{"thought": p.Thought})
		} else if p.FunctionCall != nil {
			out = append(out, map[string]any{
				"functionCall": map[string]any{
					"name": p.FunctionCall.Name,
					"id":   p.FunctionCall.Id,
					"args": p.FunctionCall.Args,
				},
			})
		}
	}
	return out
}

func functionResponsesToHistory(parts []*genai.Part) []any {
	var out []any
	for _, p := range parts {
		if p.FunctionResponse != nil {
			out = append(out, map[string]any{
				"functionResponse": map[string]any{
					"name":     p.FunctionResponse.Name,
					"id":       p.FunctionResponse.Id,
					"response": p.FunctionResponse.Response,
				},
			})
		}
	}
	return out
}

func historyToContents(history []any) []*genai.Content {
	var contents []*genai.Content
	for _, h := range history {
		m, ok := h.(map[string]any)
		if !ok {
			continue
		}
		role, _ := m["role"].(string)
		parts, _ := m["parts"].([]any)
		if role == "" || len(parts) == 0 {
			continue
		}
		var genParts []*genai.Part
		for _, p := range parts {
			pm, ok := p.(map[string]any)
			if !ok {
				continue
			}
			if text, ok := pm["text"].(string); ok && text != "" {
				genParts = append(genParts, &genai.Part{Text: text})
			} else if thought, ok := pm["thought"].(string); ok && thought != "" {
				genParts = append(genParts, &genai.Part{Thought: thought})
			} else if fc, ok := pm["functionCall"].(map[string]any); ok {
				name, _ := fc["name"].(string)
				id, _ := fc["id"].(string)
				var args json.RawMessage
				if a, ok := fc["args"]; ok {
					args, _ = json.Marshal(a)
				}
				genParts = append(genParts, &genai.Part{
					FunctionCall: &genai.FunctionCall{Name: name, Id: id, Args: args},
				})
			} else if fr, ok := pm["functionResponse"].(map[string]any); ok {
				name, _ := fr["name"].(string)
				id, _ := fr["id"].(string)
				resp := fr["response"]
				genParts = append(genParts, &genai.Part{
					FunctionResponse: &genai.FunctionResponse{Name: name, Id: id, Response: resp},
				})
			}
		}
		if len(genParts) > 0 {
			contents = append(contents, &genai.Content{Role: role, Parts: genParts})
		}
	}
	return contents
}

func thoughtSummary(raw string, maxLen int) string {
	if maxLen <= 0 {
		maxLen = 120
	}
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return ""
	}
	lines := strings.SplitN(trimmed, "\n", 2)
	first := strings.TrimSpace(lines[0])
	if len(first) > maxLen {
		return first[:maxLen] + "..."
	}
	return first
}

func thinkingStep(id, agentID, agentName, content, status, details string) map[string]any {
	m := map[string]any{
		"id":        id,
		"agentId":   agentID,
		"agentName": agentName,
		"content":   content,
		"status":   status,
		"timestamp": time.Now().UnixMilli(),
	}
	if details != "" {
		m["details"] = details
	}
	return m
}

func appendOrUpdate(steps []map[string]any, id string, step map[string]any) []map[string]any {
	for i := range steps {
		if sid, _ := steps[i]["id"].(string); sid == id {
			steps[i] = step
			return steps
		}
	}
	return append(steps, step)
}

func toolLabel(name string) string {
	defs := builtin.Definitions()
	for _, d := range defs {
		if d.Name == name && d.Description != "" {
			first := d.Description
			if idx := strings.IndexAny(first, "。；.;"); idx >= 0 {
				first = first[:idx]
			}
			first = strings.TrimSpace(first)
			if len(first) > 36 {
				first = first[:33] + "..."
			}
			return first
		}
	}
	return name
}

func sliceAny(in []any) []any {
	if in == nil {
		return nil
	}
	out := make([]any, len(in))
	copy(out, in)
	return out
}
